### ISSUE TITLE
Current stance on NA implementation

### ISSUE THREAD

**Issue by FelixBenning** [numpy/numpy#15858]
Some time ago, I actually got my statistics professor to try out python, and he came back claiming that python would never replace R due to the lack of support for NAs. 

Now I actually got around to reading up on what he meant by that. 

It appears that:
-
1. R has a bitpattern encoding of NA (missing value), distinct from NaN (not a number). 
2. There seemed to have been a discussion in the numpy community, how to implement NA (e.g. [NEP 12](https://numpy.org/neps/nep-0012-missing-data.html), [NEP 24](https://numpy.org/neps/nep-0024-missing-data-2.html), [NEP 25](https://numpy.org/neps/nep-0025-missing-data-3.html) and subsequently a summary: [NEP 26](https://numpy.org/neps/nep-0026-missing-data-summary.html) containing dead links to more threads) in 2012. 
3. There appear to be two possible implementation strategies:
   - Masks (a second table indicating whether or not an entry is missing): 
       - PRO:  Allows even more types (e.g. ignored, degree of missingness[?] apparently that is a thing, etc.) and simulating missing numbers (i.e. overlaying different missing number patterns over existing data)
   - Bitpattern encoding (This is the strategy R uses): 
        - PRO: in general faster and less memory hungry except for the case, where you want to simulate missing numbers. 
        - CON: While floats, bools and Strings/characters have spare bit patterns to indicate missingness, integers would have to sacrifice one number. 
4. The current state: 
    - Currently there seems to be a mask implementation separate from numpy core (numpy.ma) which does not appear to be widely used (e.g. frameworks like [panda  abstain and use NaNs in place of actual NAs,](https://books.google.de/books?id=xYmNDQAAQBAJ&pg=PA120&lpg=PA120&dq=numpy+bitpattern+na&source=bl&ots=XqfNnZlj7J&sig=ACfU3U0iNsAMtONgkHRlqVjh54Q3CWFIug&hl=en&sa=X&ved=2ahUKEwiuleKbqL3oAhUt2aYKHZN7Ch0Q6AEwA3oECAkQAQ#v=onepage&q=numpy%20bitpattern%20na&f=false)  this means they cast the entire dataframe to float when an integer becomes na), but no bitpattern implementation.
    - All the NEPs from above are in the "Deferred and Superseded NEPs" state on the [roadmap](https://numpy.org/neps/index.html) and I can not find any sources on the reason why. In general the discussion appears to have died down. 

I wonder
-
1. what has happened to this discussion? What was the resolution? Why does it appear that nothing came of it?
2. are the differences between na and nan so small, that nan is sufficient in most circumstances? One example where nan behaves different to na should: `np.any([False, np.nan])` results in `True`. Are there more examples?
3. are 8-bit integers really used enough to warrant the worry about losing one number for the bitpattern implementation?

My current (uneducated) opinion:
- 
If this issue prevents people from switching from R to Python, and the only real issue with bitpatterns seems to be 8-bit integers in some sense, then I do not get why this was not implemented a while ago. I doubt that there are many applications for 8-bit integers, and those could probably use a dedicated option which turns off NAs. It certainly does not seem to me, like an issue which should prevent numpy core from implementing this. That is, I might grossly underestimate the effort this would take. I have just started reading up on this issue today. 

I would hate, if another [multiple dispatch language](https://docs.julialang.org/en/v1/manual/missing/) would replace R, due to this missing feature, since multiple dispatch is one of the reasons I despise R in the first place.

**Comment by seberg** [comment]
What you will be reading about most is the bitmask style approach, that for example is now also taken by pandas and many others. Now 2012 (or even earlier) that was an approach being considered in NumPy, and it was finally stopped NEP 26 being the only definite document probably. If you unearth more information or links, improving that NEP is important! It is an informational NEP, and there is nothing wrong with ammending it.

The second idea is the idea of Bitpattern encoded missing data (as R uses it). This is simpler in many ways, and we *could* largely do it (although you still need some special hooks, e.g. at the very least you need a `count_masked` function to be used within things like `mean`.
However, we do not really believe doing this type of thing *within* NumPy, it would just slow down development. Thus, such things should preferably be written outside of NumPy. Right now datatypes written outside of NumPy have some downsides. This is not a catastrophe: You should be able to get quite far if you want to. And it is something that we will hopefully improve a lot, at which point the approach changes should however not be all lost.

So, the simple thing is: Yes, this type of thing should be approached and if NumPy needs more infrastructure to allow you to do it, we should add that (and we are working on it, although more momentum is always better).
But, whatever the approach is, bitmasks or bitpatterns: These projects will probably have to live or at least be started outside of NumPy proper. That should not be a big issue, however! For example making a new `MaskedArray` class using bitmasks (or bytemasks for simplicity...) which is better than the current `np.ma.MaskedArray` should be possible by using the new `__array_ufunc__` and `__array_function__` protocol for seamless interoperability with NumPy arrays.

The main thing we need is momentum, and that unfortunately typically means having one or more champions to push things forward. There is a big interest in this, so no matter what project someone starts on, I am sure we can ping a few people to give guidance and review!

**Comment by charris** [comment]
Just to add a bit of history, the bit pattern approach came out of Bell Labs, as did S.

**Comment by FelixBenning** [comment]
> The second idea is the idea of Bitpattern encoded missing data (as R uses it). This is simpler in many ways, and we could largely do it (although you still need some special hooks, e.g. at the very least you need a `count_masked` function to be used within things like `mean`

Simpler than what? Masks? 
In the case of floats, you would not really have to modify `mean` at all, since NA would be a special case of NaN, which the current mean function already has to deal with. Actually, scratch that, apparently there is no ignore nan option in [numpy.mean](https://docs.scipy.org/doc/numpy/reference/generated/numpy.mean.html). I guess pandas implemented that independently. In case of integers, you would have to figure out a way to implement a sum. I guess you will not get around having to make a bitpattern check for every summand. Although if we are lucky, this does not change the number of instruction cycles. I think I heard something about different types of operations being able to run in parallel on the compute unit.

> However, we do not really believe doing this type of thing within NumPy, it would just slow down development. Thus, such things should preferably be written outside of NumPy.

In the sense off create a working fork and then merge? or in the sense of, create a submodule like np.ma? Since np.ma was not very successful I doubt another attempt would be much more successful. Defaults matter. If you are not interested in making this a core functionality, few people are probably going to look for it. I mean people already complain, that you can not actually do statistics on python natively and need to import libraries like numpy for it. If you then have to also import a submodule to have NA support...

> That should not be a big issue, however! For example making a new MaskedArray class using bitmasks (or bytemasks for simplicity...) which is better than the current np.ma.MaskedArray should be possible

I am not too interested into the mask approach. The most likely case are NAs, everything else seems very niche, so I don't see why you should sacrifice performance for the additional flexibility that masks provide. I also have no idea what `ufuncs` are to be honest. I'll read up on that a bit. 

It seems like a starting point could be to try and implement a "ignore NaNs" for the mean function.

**Comment by seberg** [comment]
@FelixBenning in the sense of `np.ma`, yes `np.ma` was not very successful, but we have much better interoperability now as I mentioned `__array_function__` and `__array_ufunc__`. Sure it may still not be perfect but it will be *much* better and gives us a reason to improve the interoperability even more. `np.ma` even predates all of that, probably even the `where` argument to ufuncs probably.

Bitpatterns, you will have to create a new set of dtypes with its own universal functions, Warren has an example for that in https://github.com/WarrenWeckesser/numtypes. We may still need hooks for `np.mean` to be aware of such `dtypes` in some form (as I said), but it will be much less.

As to performance, obviously anything will typically have a performance impact, no idea how big it is in practice though.

> I am not too interested into the mask approach.

Masks are simpler in the sense that you can reuse most functionality easier, bitpatterns are maybe simpler in the sense that they do not touch the container as much (you do not have a second array). So making dtypes with bitpatterns work super-seamlessly may actually be more likely in a mid-long term, but bitmasks are probably *much* more doable in the short-mid term. And we may be able to learn a lot from a good bitmask implementation (e.g. what functionality would you want). 

You can implement an `ignore NaN` in mean, but we already have that, it is called `np.nanmean`. I think if we had `NaN` bitpattern types, we may need to check and by default ignore them in `np.mean`, my reasoning would be that `NaN` is normally propagating, while a masked value is normally an ignore value. An integer dtype may want to support both types (or are there more?) if it was created.

**Comment by FelixBenning** [comment]
I have to revise my statement. It appears that Masks might not be less performant after all:
Wes McKinney argues here that bitmaps are actually more performant
https://wesmckinney.com/blog/bitmaps-vs-sentinel-values/
And this benchmark follow up by Pasha Stetsenko shows, that they can be made roughly equivalent:
https://github.com/st-pasha/microbench-nas/blob/master/README.md

Wes further argues that for bitmaps: "Many
computational operations (like elementwise functions) need not concern
themselves with the nulls at all, for example, since the bitmap from
the input array can be passed along (with zero copy even) to the
output array."

My own thoughts: It might also be interesting to explicitly have zeros in the array where the bitmap indicates an NA, since a sum ignoring NAs can literally just ignore the fact that NAs exist, and sum over the entire colum. Similarly a scalar product would simply cause the products to be zero (i.e. ignored) where either value is zero. So the scalar product could just be calculated as if there was no NAs. This increases the costs of operations, which would result in non-zero fields where NAs should be, which would then have to be updated.
It might make sense to set all NAs to zero only before a costly operation like matrix multiplication, since that only costs O(n), while the matrix multiplication is vastly more expensive. Maybe a flag like na_are_zero would make sense for this.

**Comment by FelixBenning** [comment]
#15917

**Comment by seberg** [comment]
To be honest, I am not sure that discussions about what is more performant is interesting here. What might be interesting is points along the lines of external implementation being able to use `__array_function__`, `__array_ufunc__`, etc. or being able to use bitpattern approach based on implementing new dtypes.

To be clear: it is unlikely that NA handling can land in NumPy without first being developed outside of it at this point. We are happy to facilitate that of course, and I *do* think that we should adapt NumPy to achieve that where necessary. We should clarify and maybe fix links, but I do not think we have to try and continue the discussion of the best approach in these NEPs.

To be honest, there must be a huge amount of prior art along similar lines to what you linked, and I am unwilling to judge whether linking this in the NEP helps the reader.

**Comment by FelixBenning** [comment]
Well I only tried to update the NEP with the new information I found as you suggested. You may do what you want with it. Which is also why I only changed the references and not anything in the document itself.

**Comment by seberg** [comment]
Sorry yeah, thanks for the suggestion, maybe we can put it in in any case. It seemed to me that you felt the current document is unclear as to the actual current situation, so that was what I was mainly thinking of.
But maybe we can actually extend it in this direction as well.

**Comment by rossbar** [comment]
It seems that the original discussion (bitmasks vs. masks, esp. from the perspective of performance) has been addressed by some subsequent research. It was decided not to add new references to NEP 26, mostly because it didn't make a lot of sense to drop-in references without any of the context from supporting discussion.

I also gather from the above discussion that the better path to better support for masking would be to improve integration/interoperability with `np.ma`. To me, it seems that this would be the more productive discussion, so I'm inclined to close this one. It can be linked to from any relevant MA discussions to provide a little background on the different approaches to masking values.
